File: backend/src/models/message_analysis.ts
import mongoose, { Schema, Document } from 'mongoose';

/**
 * Message Analysis Model
 * Task: T002 - Create MessageAnalysis model schema
 * Feature: 002-intelligent-message-analysis
 */

export type SentimentType = 'positive' | 'neutral' | 'negative';

export interface IMessageAnalysis extends Document {
  messageId: mongoose.Types.ObjectId;
  analysisTimestamp: Date;
  spamProbability: number;  // 0-100%
  isSpam: boolean;
  needsResponse: boolean;
  responseConfidence: number;  // 0-100%
  sentiment: SentimentType;
  priorityLevel: 'high' | 'medium' | 'low';
  generatedReplyText?: string;
  analysisVersion: string;
  createdAt: Date;
  updatedAt: Date;

  // Methods
  shouldAutoDelete(threshold?: number): boolean;
  shouldAutoReply(threshold?: number): boolean;
}

const MessageAnalysisSchema = new Schema<IMessageAnalysis>({
  messageId: {
    type: Schema.Types.ObjectId,
    ref: 'Message',
    required: true,
    unique: true,  // One analysis per message
    index: true
  },
  analysisTimestamp: {
    type: Date,
    default: Date.now,
    required: true
  },
  spamProbability: {
    type: Number,
    required: true,
    min: 0,
    max: 100,
    index: true
  },
  isSpam: {
    type: Boolean,
    required: true,
    default: false,
    index: true
  },
  needsResponse: {
    type: Boolean,
    required: true,
    default: false,
    index: true
  },
  responseConfidence: {
    type: Number,
    required: true,
    min: 0,
    max: 100,
    index: true
  },
  sentiment: {
    type: String,
    enum: ['positive', 'neutral', 'negative'],
    required: true,
    default: 'neutral',
    index: true
  },
  priorityLevel: {
    type: String,
    enum: ['high', 'medium', 'low'],
    required: true,
    default: 'medium',
    index: true
  },
  generatedReplyText: {
    type: String
  },
  analysisVersion: {
    type: String,
    required: true,
    default: '1.0'
  }
}, {
  timestamps: true,
  collection: 'message_analyses'
});

// Compound indexes for common queries
MessageAnalysisSchema.index({ isSpam: 1, spamProbability: -1 });
MessageAnalysisSchema.index({ needsResponse: 1, responseConfidence: -1 });
MessageAnalysisSchema.index({ sentiment: 1, priorityLevel: 1 });

// Instance methods
MessageAnalysisSchema.methods.shouldAutoDelete = function(threshold: number = 80): boolean {
  return this.isSpam && this.spamProbability >= threshold;
};

MessageAnalysisSchema.methods.shouldAutoReply = function(threshold: number = 85): boolean {
  return this.needsResponse && this.responseConfidence >= threshold;
};

// Pre-save hook to set isSpam based on threshold
MessageAnalysisSchema.pre('save', function(next) {
  if (this.isNew || this.isModified('spamProbability')) {
    this.isSpam = this.spamProbability >= 80;
  }
  next();
});

export const MessageAnalysis = mongoose.model<IMessageAnalysis>('MessageAnalysis', MessageAnalysisSchema);

File: flutter_app/lib/models/message_analysis.dart
import 'package:freezed_annotation/freezed_annotation.dart';
import 'message.dart';

part 'message_analysis.freezed.dart';
part 'message_analysis.g.dart';

/// Message Analysis Model
/// Task: T029 - Create MessageAnalysis Freezed model
/// Reference: specs/002-intelligent-message-analysis/plan.md

enum Sentiment {
  positive,
  neutral,
  negative,
}

@freezed
class MessageAnalysis with _$MessageAnalysis {
  const MessageAnalysis._();

  const factory MessageAnalysis({
    required String id,
    required String messageId,
    required DateTime analysisTimestamp,
    @Default(0.0) double spamProbability,
    @Default(false) bool isSpam,
    @Default(false) bool needsResponse,
    @Default(0.0) double responseConfidence,
    @Default(Sentiment.neutral) Sentiment sentiment,
    @Default(PriorityLevel.medium) PriorityLevel priorityLevel,
    String? generatedReplyText,
    @Default('1.0') String analysisVersion,
  }) = _MessageAnalysis;

  factory MessageAnalysis.fromJson(Map<String, dynamic> json) =>
      _$MessageAnalysisFromJson(json);

  // Computed properties
  bool get isLikelySpam => spamProbability > 70.0;
  bool get isDefinitelySpam => spamProbability > 90.0;
  bool get shouldGenerateReply => needsResponse && responseConfidence > 60.0;
  bool get hasGeneratedReply => generatedReplyText != null && generatedReplyText!.isNotEmpty;

  String get sentimentLabel {
    switch (sentiment) {
      case Sentiment.positive:
        return 'Positif';
      case Sentiment.neutral:
        return 'Neutre';
      case Sentiment.negative:
        return 'Négatif';
    }
  }

  String get spamLabel {
    if (isDefinitelySpam) return 'Spam confirmé';
    if (isLikelySpam) return 'Spam probable';
    if (spamProbability > 40.0) return 'Spam possible';
    return 'Pas de spam';
  }
}
